lex/outputbis.ml

```diff
@@ -126,97 +165,148 @@ let output_moves oc has_refill moves =
     (fun i -> match i with
-    | SetTag (t,m) -> fprintf oc " t%d <- [%d] ;" t m
-    | EraseTag t -> fprintf oc " t%d <- -1 ;" t)
+    | SetTag (t,m) -> pr ctx " t%d <- [%d] ;" t m
+    | EraseTag t -> pr ctx " t%d <- -1 ;" t)
     mvs ;
-  output_string oc " *)\n" ;
+  pr ctx " *)\n" ;
   List.iter
     (fun i ->  match i with
     | SetTag (t,m) ->
-        fprintf oc "%s%a <- %a ;\n"
+        pr ctx "%s%a <- %a ;\n"
           pref output_mem_access t output_mem_access m
     | EraseTag t ->
-        fprintf oc "%s%a <- -1 ;\n"
+        pr ctx "%s%a <- -1 ;\n"
           pref output_mem_access t)
     mvs
 
-let output_trans pref oc has_refill i trans =
-  let entry = sprintf "__ocaml_lex_state%d" i in
-  fprintf oc "%s %s lexbuf %s= " pref entry
-    (if has_refill then "k " else "");
-  match trans with
+let output_trans_body ctx = function
   | Perform (n,mvs) ->
-      output_tag_actions "  " oc mvs ;
-      fprintf oc "  %s%d\n"
-        (if has_refill then "k lexbuf " else "")
+      output_tag_actions "  " ctx mvs ;
+      pr ctx  {|
+  lexbuf.Lexing.lex_curr_pos <- _curr;
+  lexbuf.Lexing.lex_last_pos <- _last;
+|};
+      pr ctx "  %s%d\n"
+        (if ctx.has_refill then Printf.sprintf "k lexbuf " else "")
         n
   | Shift (trans, move) ->
-    begin match trans with
-      | Remember (n,mvs) ->
-        output_tag_actions "  " oc mvs ;
-        fprintf oc
-          "  lexbuf.Lexing.lex_last_pos <- lexbuf.Lexing.lex_curr_pos ;\n" ;
-        fprintf oc "  lexbuf.Lexing.lex_last_action <- %d ;\n" n;
-      | No_remember -> ()
-    end;
-    if has_refill then
-      let next = entry ^ "_next" in
-      fprintf oc "  __ocaml_lex_next_char lexbuf %s k\n" next;
-      fprintf oc "and %s lexbuf k = function " next
-    else
-      output_string oc "match __ocaml_lex_next_char lexbuf with\n";
-    output_moves oc has_refill move
+      let ctx =
+        match trans with
+        | Remember (n,mvs) ->
+            output_tag_actions "  " ctx mvs ;
+            pr ctx
+              "  let _last = _curr in\n";
+            begin match ctx.last_action with
+            | Some i when i = n ->
+                pr ctx "  (* let _last_action =  %d in*)\n" n;
+                ctx
+            | _ ->
+                pr ctx "  let _last_action = %d in\n" n;
+                {ctx with last_action = Some n}
+            end
+        | No_remember ->
+            ctx
+      in
+      if ctx.has_refill then begin
+        (* TODO: bind this 'state' function at toplevel instead *)
+        pr ctx {|
+  let state lexbuf _last_action _buf _len _curr _last k = function
+|};
+      output_moves ctx move;
+      pr ctx {|
+  in
+  if _curr >= _len then
+     __ocaml_lex_refill_buf lexbuf _buf _len _curr _last _last_action state k
+  else
+     state lexbuf _last_action _buf _len (_curr + 1) _last k
+       (Char.code (Bytes.unsafe_get _buf _curr))
+|}
+      end
+      else begin
+        pr ctx {|
+  let next_char, _buf, _len, _curr, _last =
```

Comment:
This depends critically on the optimized compilation scheme for tuple-binding. We would not want to allocate the 4-tuple for every character.